### PR TITLE
pyln-testing: Fix BitcoinD FD leak

### LINKED ISSUE #7130: pyln-testing: BitcoinD TailableProc leaking file descriptors
### Issue and Steps to Reproduce
Running pytest locally leaks file descriptors for TailableProc `stdout`, read/write handles opened in the `__init__()` constructor. This resulted in sporadic failures and was effectively addressed by raising the `ulimit`. Current testing strategy is:

1. `make pytest`
2. In a 2nd terminal window get the process idea with `ps aux | grep pytest`
3. In a 3rd terminal window, observe the number of opened files grow with `watch -n 1 'lsof -p ${PID} | wc`
4. In a 4th terminal window, periodically run `lsof -p ${PID} | less +G` to see the file descriptors grow for each test.

`lsof` reports four file descriptors hanging around for each test which are initialized in the `TailableProc` constructor [here](https://github.com/ElementsProject/lightning/blob/d1101f416f1ec959981a8ebe5639a6b267885bfd/contrib/pyln-testing/pyln/testing/utils.py#L203).

### `getinfo` output

N/A

### PR DESCRIPTION
**Pull request by s373nZ**
# Fix BitcoinD FD leak in pyln-testing

## Description
This PR adds a post-yield resolver sanity-check to the `bitcoind` test fixture to be sure file descriptors are not leaking.

## Related Issues
- Closes #7130 (if it works)
- I made this PR with the faint hope that it might be related to the build failures in #7310 as well. Need to trigger CI to see...

## Changes Made
- [ ] **Feature**: Brief description of the new feature or functionality added.
- [x] **Bug Fix**: Brief description of the bug fixed and how it was resolved.
- [ ] **Refactor**: Any code improvements or refactoring done without changing the functionality.

## Checklist
Ensure the following tasks are completed before submitting the PR:

- [x] Changelog has been added in relevant commit/s.
- [ ] Tests have been added or updated to cover the changes.
- [ ] Documentation has been updated as needed.
- [ ] Any relevant comments or `TODOs` have been addressed or removed.

## Additional Notes
Please let me know if there is a more elegant way to do this, or more appropriate place to put it.

### CODE REVIEW

**Comment by s373nZ** [comment]
Refactored the file closing operations into a function on `TailableProc` and call it from the `bitcoind` fixture.

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,11 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## [Unreleased]
+
+### Added
+
+- pyln-testing: Fix FD leaks in BitcoinD. ([#7130])
```
Comment (ShahanaFarooqui) [inline CHANGELOG.md:10]: Changelog needs to be added in the commit not in the file as explained https://docs.corelightning.org/docs/coding-style-guidelines#changelog-entries-in-commit-messages.

**Review by ShahanaFarooqui — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (s373nZ) [reply CHANGELOG.md]: Thanks! Just learned about this :) Fixed here and will address in my other PRs as well.

**Comment by ShahanaFarooqui** [comment]
@s373nZ Sorry if I overlooked it while jumping between 7669, 7108, 7130, and 7310 :). 
Could you please remind why the cleanup process is not added to the [finalizer](https://github.com/ElementsProject/lightning/issues/7130#issuecomment-1983465331) or the [stop](https://github.com/ElementsProject/lightning/blob/d1101f416f1ec959981a8ebe5639a6b267885bfd/contrib/pyln-testing/pyln/testing/utils.py#L242) function?

I am testing it with:
- 1st terminal: `make pytest`
- 2nd terminal: Get the process id with `ps aux | grep pytest`
- 3rd terminal: `lsof -p ${PID} | wc -l`

And it seems to be working (tests are still running at 32% :-|).

**Review by ShahanaFarooqui — APPROVED** [review]

**Comment by s373nZ** [comment]
@ShahanaFarooqui That's a lot of jumping (and a lot of 7s :). I should have done a better job at consolidating this information into the PR description.

### Context

For some context, I ran into this file descriptor issue a few months ago when working on a PR and attempting to run the `make pytest` locally. The leak produced seemingly random test failures that weren't alleviated until I lifted the `ulimit`. New to CLN development, I was troubleshooting the issue then under the assumption that the `LightningD` subprocess was responsible, and tried a wide variety of fixes. Returning to the issue recently in hopes that a fix might stabilize the CI tests somewhat, fresh eyes revealed it is actually the `BitcoinD` process.

To address your questions directly:

> could you please remind why the cleanup process is not added to the https://github.com/ElementsProject/lightning/issues/7130#issuecomment-1983465331

My reading of the Pytest documentation suggested by @cdecker in that comment, specifically the preceding section titled [Teardown/Cleanup (AKA Fixture finalization)](https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended), suggests that `yield` fixtures are recommended and _Any teardown code for that fixture is placed after the yield_. The `bitcoind` fixture is already implemented as a `yield` fixture, so closing the files as a cleanup operation performed after the `yield` statement in the current implementation aligns with best practice recommendations as I understand it.

The following section in the docs, [Adding finalizers directly](https://docs.pytest.org/en/6.2.x/fixture.html#adding-finalizers-directly) reads to apply typically to non-yield fixtures. Seeing as this is actually an issue with *every* test case which leverages the `bitcoind` fixture (each leaves four fds open in a `(deleted)` state throughout the entire `pytest` suite run), I understand the file closing operations to be necessarily bundled with the yield fixture. I think at the time, @cdecker and I thought that the leak might have affected a subset of test cases and were exploring `addfinalizer()` as a targeted solution.

In short, the implementation in this PR is indeed using finalizers, but in accordance with yield fixtures rather than `addfinalizer`.

> ...or the [stop](https://github.com/ElementsProject/lightning/blob/d1101f416f1ec959981a8ebe5639a6b267885bfd/contrib/pyln-testing/pyln/testing/utils.py#L242) function?

This is a good question, and one I tried to implement initially. `TailableProc` is inherited by *both* [`BitcoinD`](https://github.com/s373nZ/lightning/blob/e35e3c540b60fab1a626ca9df1b6d07225c154c0/contrib/pyln-testing/pyln/testing/utils.py#L394) and [`LightningD`](https://github.com/s373nZ/lightning/blob/e35e3c540b60fab1a626ca9df1b6d07225c154c0/contrib/pyln-testing/pyln/testing/utils.py#L590). There are a number of existing tests which `stop()` and `start()` the Lightning nodes, and continue to make use of the files throughout the tests to validate log messages. If we close these files when the daemon is stopped, these tests break. Additionally, there are a number of global "[teardown checks](https://github.com/s373nZ/lightning/blob/e35e3c540b60fab1a626ca9df1b6d07225c154c0/contrib/pyln-testing/pyln/testing/fixtures.py#L506)" in the `node_factory` fixture itself which depend on the log files to remain open. After trying a wide variety of implementations using `stop()` and even moving the file `open()` calls out of the initializer into `start()`, it became apparent that this approach was too invasive with the current test suite.

Now that it's discovered `BitcoinD` is the leak culprit, there is still a small mystery as to why `LightningD` isn't leaking file descriptors in the same manner. My best understanding at the moment is that its daemon is started in a subprocess using `Popen()` [here](https://github.com/s373nZ/lightning/blob/e35e3c540b60fab1a626ca9df1b6d07225c154c0/contrib/pyln-testing/pyln/testing/utils.py#L236), and according to the [docs](https://docs.python.org/3/library/subprocess.html#popen-constructor):
> If close_fds is true, all file descriptors except 0, 1 and 2 will be closed before the child process is executed. Otherwise when close_fds is false, file descriptors obey their inheritable flag as described in [Inheritance of File Descriptors](https://docs.python.org/3/library/os.html#fd-inheritance).
>
> Changed in version 3.2: The default for close_fds was changed from [False](https://docs.python.org/3/library/constants.html#False) to what is described above.

So, I believe the subprocess is taking care of closing `LightningD`'s file descriptors by default, but `BitcoinD` doesn't get the same treatment.

### Alternative implementations

- While closing the files in `stop()` doesn't seem to be the right solution, there may be other options. Another idea was to implement it in the `TailableProc` destructor `__del__()`, but I'm pretty sure I tried that last March and didn't get it working.

- Also, I considered the name `cleanup()` for the callback function in `TailableProc`, which is a more general name. However, that function would be implicitly overridden by `LightningD` [here](https://github.com/s373nZ/lightning/blob/e35e3c540b60fab1a626ca9df1b6d07225c154c0/contrib/pyln-testing/pyln/testing/utils.py#L648) when called by `LightningNode` [here](https://github.com/s373nZ/lightning/blob/e35e3c540b60fab1a626ca9df1b6d07225c154c0/contrib/pyln-testing/pyln/testing/utils.py#L648). This could probably work, too, but it was preferable to avoid unexpected behavior by injecting assumptions into the existing inheritance hierarchy.

Hope this helps! I haven't mastered Pytest, Pyln, or the CLN test suite by any means, so please let me know if you have recommendations for a better fit solution.

**Comment by s373nZ** [comment]
Also, I have one tangential question, if you would. When referencing issues in the git commit messages, I've been using the Github issue number that describes the problem both in the first line of the commit message and in the `Changelog-*: ` value. Should the number reference the PR instead? If so, let me know and I'll amend my commit messages.

**Comment by ShahanaFarooqui** [comment]
> Also, I have one tangential question, if you would. When referencing issues in the git commit messages, I've been using the Github issue number that describes the problem both in the first line of the commit message and in the `Changelog-*: ` value. Should the number reference the PR instead? If so, let me know and I'll amend my commit messages.

Typically, including the issue number in commit messages is useful for tracking purposes. It is recommended to reference issues using `related`, `fixes`, or `closes` in the PR description as well. 

#### CHANGELOG list update:
Before each release, the [changelog.py](https://github.com/ElementsProject/lightning/blob/master/devtools/changelog.py) script generates entries based on commits merged after the last tag.

For instance, the script will produce below entry for your commit message, "Changelog-Fixed: pyln-testing: Fix file descriptor leak in bitcoind fixture.":

```
...
### Fixed
 - pyln-testing: Fix file descriptor leak in bitcoind fixture. ([#7669])

... 
[#7669]: https://github.com/ElementsProject/lightning/pull/7669
```

This way, your changes are automatically included in the appropriate section of the changelog.

**Review by ShahanaFarooqui — APPROVED** [review]

**Comment by ShahanaFarooqui** [comment]
ACK e35e3c540b60fab1a626ca9df1b6d07225c154c0.

**Comment by ShahanaFarooqui** [comment]
I will merge the PR after final review from @cdecker.

**Comment by whitslack** [comment]
> Now that it's discovered `BitcoinD` is the leak culprit, there is still a small mystery as to why `LightningD` isn't leaking file descriptors in the same manner.

Actually, `LightningD` _does_ leak file descriptors in the same manner as `BitcoinD` did before this fix went in, at least on my machine. I was seeing four dangling file descriptors for each and every `LightningD` instance started, which would never be cleaned up and so would lead to file descriptor exhaustion. I have fixed that in #8488 by looping through `nf.nodes` at the end of the `node_factory` fixture generator and calling `cleanup_files()` on each node.

**Pull request merged**